enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,348 @@
+   See the `Upgrading`section.
+
+### Test Plan
+
+We will run additional test matrix with `microshift-core` alone, in those
+tests any OVN/NetworkPolicy tests would need to be disabled.
+
+Additional unit tests should cover the loading of core components as
+kustomize manifests, as well as the depencency handing between them.
+
+Additional unit tests should cover the loading of multiple `.d` style
+directories with manifests.
+
+## Implementation History
+
+* [openshift/microshift](https://github.com/openshift/microshift)
+* [Design guidelines](https://github.com/openshift/microshift/blob/main/docs/design.md)
+
+## Alternatives
+
+### Using go plugins which microshift can load when available, intead of yamls.
+
+[go/plugin](https://pkg.go.dev/plugin) could be used to create an interface
+for loadable components, MicroShift would scan an spefic directory for .so files
+
+Every plugin would contain it's own list of images, setup functionality,
+validation hooks, etc.
+
+The Risks here are:
+
+* Binary code duplication: How do we handle shared code between MicroShift
+  and the plugins, for example plugins would be expected to call kustomize
+  libraries, but can't call back into code in MicroShift binary. This would
+  force us replicate those libraries again statically in the plugins along
+  with the kubernetes clients
+
+* The inteface and calling convention needs to be maintained with versioning
+  to avoid externally created plugins from breaking as we evolve the API.
+  This is also true for yaml based components, but it's simpler to provide
+  additive iterations without breaking backwards compatibility in that case.
+  
+* CGO_ENABLED needs to be enabled.
+
+* Requires same version of go vendor dependencies.
+
+The possible benefits are:
+
+* Easier to handle validations of the system
+* More flexible ability to read and handle configuration files.
+* Upgrade of CRDs can potentially be done
+
+### Controlling component behaviour from configuration/go code
+
+We could add flags to the MicroShift configuration which then can control
+the component loading via flags.
+
+Image versions would be switched at compile time for upstream versions
+where necessary. Components which we could not ship upstream would be disabled
+out from compilation via build -tags.
+
+All the code relevant to the non-loaded componets would remain in the binary.
+
+The possible benefits are:
+
+* It's probably the simplest implementation.
+
+Risks:
+
+* It does not prevent coupling creep. Components would continue to couple with
+  the MicroShift codebase in unexpected ways, as we already see today.
+
+### External binaries
+
+MicroShift would call on the external binaries with a kubeconfig, and use a
+calling interface to request actions like setup/initialization/upgrades/etc.
+
+Risks:
+
+* Binary code duplication in disk, as the kubernetes and other related libraries
+  need to be compiled again into another binary.
+
```

Comment:
Add possible benefits:
- fully decoupled plugin logic in external binary
- external binary can run independently

Reply:
That's the same same in the current main proposal, though.